source/declarations.tex

```diff
@@ -4239,3 +4240,414 @@
+for the return value and/or the state of objects
+using a predicate that is intended to hold upon exit from the function.
+A postcondition may introduce an identifier to represent
+the glvalue result or the prvalue result object of the function.
+\begin{example}
+\begin{codeblock}
+int f(char * c)
+  [[ensures res: res > 0 && c != nullptr]];
+
+int g(double * p)
+  [[ensures audit res: res != 0 && p != nullptr && *p <= 0.0]];
+\end{codeblock}
+\end{example}
+
+\pnum
+A \grammarterm{contract-attribute-specifier} using \tcode{assert}
+is an \defn{assertion}.
+It expresses a condition that is intended to be satisfied
+where it appears in a function body.
+
+\pnum
+Preconditions, postconditions, and assertions
+are collectively called \defnx{contracts}{contract}.
+The \grammarterm{conditional-expression} in a contract
+is contextually converted to \tcode{bool}\iref{conv};
+the converted expression is called
+the \defnx{predicate}{contract!predicate} of the contract.
+
+\pnum
+A \defn{contract condition} is a precondition or a postcondition.
+A contract condition may be applied to
+the function type of a function declaration.
+The first declaration of a function shall specify
+all contract conditions (if any) of the function.
+Subsequent declarations shall either specify no contract conditions
+or the same list of contract conditions;
+no diagnostic is required
+if corresponding conditions will always evaluate to the same value.
+The list of contract conditions of a function shall be the same
+if the declarations of that function appear in different translation units;
+no diagnostic required.
+If a friend declaration
+is the first declaration of the function in a translation unit
+and has a contract condition,
+the declaration shall be a definition and
+shall be the only declaration of the function in the translation unit.
+
+\pnum
+Two lists of contract conditions are the same
+if they consist of the same contract conditions in the same order.
+Two contract conditions are the same
+if their contract levels are the same and their predicates are the same.
+Two predicates contained in \grammarterm{contract-attribute-specifier}{s}
+are the same
+if they would satisfy the one-definition rule\iref{basic.def.odr}
+were they to appear in function definitions,
+except for renaming of parameters,
+return value identifiers (if any),
+and template parameters.
+
+\pnum
+\begin{note}
+A function pointer cannot include contract conditions.
+\begin{example}
+\begin{codeblock}
+typedef int (*fpt)() [[ensures r: r != 0]];     // error: contract condition not on a function declaration
+
+int g(int x)
+  [[expects: x >= 0]]
+  [[ensures r: r > x]]
+{
+  return x+1;
+}
+
+int (*pf)(int) = g;                             // OK
+int x = pf(5);                                  // contract conditions of \tcode{g} are checked
+\end{codeblock}
+\end{example}
+\end{note}
+
+\pnum
+The predicate of a contract is potentially evaluated\iref{basic.def.odr}.
```

Comment:
Move this sentence to the end of paragraph 5. Also, we should convert it to a note, since this follows from the definition of "potentially evaluated".

Reply:
Agreed.